### DISCUSSION TITLE
X-Forwarded-For

### DISCUSSION THREAD

**Discussion by jk20004** [telekom-security/tpotce#953] [Q&A]
We are on the way to use the tpot for http traffic where we know that the requests illegal so we make the https offloading and send the data to the tpot.

The Problem is that these requests will shown with the IP from the load-balancer and not from the attacker.

is it posible to use the X-Forwarded-For Header value for the IP?

**Comment by t3chn0m4g3** [comment]
Probably not, as the web honeypot you are forwarding the request to would need to support extracting the XFF header. The best way is to route the traffic to achieve your goal and at the same time have the session working.

**Comment by BenediktS631** [comment]
Hey there,

so we have the same problem. And we want to start a more detailed discussion about it.

Our goal is to use the t-pot primarily for http traffic. 

The problem is that the requests go through a load balancer beforehand, which sets the source IP to that of the load balancer. 

To solve this problem we want to use the X-Forwarded-For header to pass the real source IP. However, if I see this correctly, this requires that the individual honeypots must support reading the X-Forwarded-For header.

And now the question: 
Would it be possible to implement such functionality for all honeypots that use http? Especially for "honeypots"([honeypots](https://github.com/qeeqbox/honeypots))

**Reply by t3chn0m4g3** [reply]
> And now the question:
Would it be possible to implement such functionality for all [honeypots](https://github.com/qeeqbox/honeypots) that use http?

If implemented upstream I will definitely add ENV options for the corresponding docker-compose files, so you can easily take advantage of the features.

**Reply by btkev123** [reply]
Hello @t3chn0m4g3 
thanks for this answer. When can we expect the possibility for ENV options, do you have a roadmap?
Thanks in advance for your support.

**Reply by t3chn0m4g3** [reply]
I was looking at the [qeeqbox/honeypots](https://github.com/qeeqbox/honeypots) (as the upstream) repository and I could not see any changes?
There was no [issue](https://github.com/qeeqbox/honeypots/issues) opened, a [PR](https://github.com/qeeqbox/honeypots/pulls?q=is%3Apr+is%3Aclosed) added or [commit](https://github.com/qeeqbox/honeypots/commits/main) pushed that requests or adds such functionality.
Only once the feature is added upstream I can take care of the rest, i.e. ENV options for the docker image.

**Reply by btkev123** [reply]
I have opened a new [issue #16](https://github.com/qeeqbox/honeypots/issues/16).
Thanks in advance :-)

**Reply by giga-a** [reply]
Hey @t3chn0m4g3 😃

I got that implemented in 0.49, there is logic that checks for `X-Forwarded-For` and `X-Real-IP`, if that does not work, it will use the loadbalancer ip (This logic is not turned on by default)

I made that logic accessible through environment variables based on your comment, you can create an environment variable named `HONEYPOTS_OPTIONS` and add `fix_get_client_ip`

Or, add `options` variable with `fix_get_client_ip` in the `http` and `https` in the tpot config file
```json
      "http":{
         "port":80,
         "ip":"0.0.0.0",
         "username":"http",
         "password":"anonymous",
         "log_file_name":"http.log",
         "max_bytes":10000,
         "backup_count":10,
         "options":"fix_get_client_ip"
      }
```

before
`
"server": "http_server", "src_ip": "0.0.0.0", "src_port": "80", "status": "success", "timestamp": "2022-02-10T20:18:59.828716", "username": "test"}
honeypots_1     | {"action": "connection", "data": {"Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8", "Accept-Encoding": "gzip, deflate", "Accept-Language": "en-US,en;q=0.5", "Connection": "close", "Host": "0.0.0.0", "Upgrade-Insecure-Requests": "1", "User-Agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0", "X-Forwarded-For": "172.20.0.1", "X-Real-Ip": "172.20.0.1", "method": "GET", "uri": "/"}, "dest_ip": "0.0.0.0", "dest_port": "80", "server": "http_server", "src_ip": "172.20.0.3", "src_port": "54420", "timestamp": "2022-02-10T20:20:14.075570"}
`

after
`
honeypots_1     | {"action": "process", "dest_ip": "0.0.0.0", "dest_port": "80", "password": "test", "server": "http_server", "src_ip": "0.0.0.0", "src_port": "80", "status": "success", "timestamp": "2022-02-10T20:29:59.802467", "username": "test"}
honeypots_1     | {"action": "connection", "data": {"Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8", "Accept-Encoding": "gzip, deflate", "Accept-Language": "en-US,en;q=0.5", "Connection": "close", "Host": "0.0.0.0", "Upgrade-Insecure-Requests": "1", "User-Agent": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:96.0) Gecko/20100101 Firefox/96.0", "X-Forwarded-For": "172.21.0.1", "X-Real-Ip": "172.21.0.1", "method": "GET", "uri": "/"}, "dest_ip": "0.0.0.0", "dest_port": "80", "server": "http_server", "src_ip": "172.21.0.1", "src_port": "34992", "timestamp": "2022-02-10T20:30:09.458675"}
`

And, I enabled mocking for some honeypots by default (http, https, ftp, imap, pop3, and ssh)

![mocking](https://user-images.githubusercontent.com/52905881/153492538-b19659fb-1e3b-4503-aa21-caf7bf11b040.png)

Please let me know if you run into any issue with this new update..

**Reply by t3chn0m4g3** [reply]
Thank you @giga-a. This will make @jk20004 @btkev123 & @BenediktS631 very happy!

**Reply by btkev123** [reply]
Yeah, thanks a lot for the quick solution!
I will check it soon.

**Reply by t3chn0m4g3** [reply]
I adjusted the configuration file (`config.json`) as suggested by @giga-a.
New docker images have been built and are available through the [development branch](https://github.com/telekom-security/tpotce/tree/22.x). For testing you can install the dev version (`makeiso.sh`), choose `mini` via `tped.sh`. Let me know if things work as expected.
